File: src/main/java/net/minestom/server/monitoring/BenchmarkManager.java
package net.minestom.server.monitoring;

import it.unimi.dsi.fastutil.longs.Long2LongMap;
import it.unimi.dsi.fastutil.longs.Long2LongOpenHashMap;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.TextComponent;
import net.kyori.adventure.text.format.NamedTextColor;
import net.minestom.server.MinecraftServer;
import net.minestom.server.utils.MathUtils;
import net.minestom.server.utils.validate.Check;
import org.jetbrains.annotations.NotNull;

import java.lang.management.ManagementFactory;
import java.lang.management.ThreadInfo;
import java.lang.management.ThreadMXBean;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import static net.minestom.server.MinecraftServer.THREAD_NAME_TICK;
import static net.minestom.server.MinecraftServer.THREAD_NAME_TICK_SCHEDULER;

/**
 * Small monitoring tools that can be used to check the current memory usage and Minestom threads CPU usage.
 * <p>
 * Needs to be enabled with {@link #enable(Duration)}. Memory can then be accessed with {@link #getUsedMemory()}
 * and the CPUs usage with {@link #getResultMap()} or {@link #getCpuMonitoringMessage()}.
 * <p>
 * Be aware that this is not the most accurate method, you should use a proper java profiler depending on your needs.
 */
public final class BenchmarkManager {
    private static final ThreadMXBean THREAD_MX_BEAN = ManagementFactory.getThreadMXBean();
    private static final List<String> THREADS = new ArrayList<>();

    static {
        THREADS.add(THREAD_NAME_TICK_SCHEDULER);
        THREADS.add(THREAD_NAME_TICK);
    }

    private final Long2LongMap lastCpuTimeMap = new Long2LongOpenHashMap();
    private final Long2LongMap lastUserTimeMap = new Long2LongOpenHashMap();
    private final Long2LongMap lastWaitedMap = new Long2LongOpenHashMap();
    private final Long2LongMap lastBlockedMap = new Long2LongOpenHashMap();
    private final Map<String, ThreadResult> resultMap = new ConcurrentHashMap<>();

    private boolean enabled = false;
    private volatile boolean stop = false;
    private long time;

    public void enable(@NotNull Duration duration) {
        Check.stateCondition(enabled, "A benchmark is already running, please disable it first.");
        THREAD_MX_BEAN.setThreadContentionMonitoringEnabled(true);
        THREAD_MX_BEAN.setThreadCpuTimeEnabled(true);

        this.time = duration.toMillis();

        final Thread thread = new Thread(null, () -> {
            while (!stop) {
                refreshData();
                try {
                    Thread.sleep(time);
                } catch (InterruptedException e) {
                    MinecraftServer.getExceptionManager().handleException(e);
                }
            }
            stop = false;
        }, MinecraftServer.THREAD_NAME_BENCHMARK);
        thread.setDaemon(true);
        thread.start();

        this.enabled = true;
    }

    public void disable() {
        this.stop = true;
        this.enabled = false;
    }

    public void addThreadMonitor(@NotNull String threadName) {
        THREADS.add(threadName);
    }

    /**
     * Gets the heap memory used by the server in bytes.
     *
     * @return the memory used by the server
     */
    public long getUsedMemory() {
        return Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
    }

    public @NotNull Map<String, ThreadResult> getResultMap() {
        return Collections.unmodifiableMap(resultMap);
    }

    public @NotNull Component getCpuMonitoringMessage() {
        Check.stateCondition(!enabled, "CPU monitoring is only possible when the benchmark manager is enabled.");
        TextComponent.Builder benchmarkMessage = Component.text();
        for (var resultEntry : resultMap.entrySet()) {
            final String name = resultEntry.getKey();
            final ThreadResult result = resultEntry.getValue();

            benchmarkMessage.append(Component.text(name, NamedTextColor.GRAY));
            benchmarkMessage.append(Component.text(": "));
            benchmarkMessage.append(Component.text(MathUtils.round(result.getCpuPercentage(), 2), NamedTextColor.YELLOW));
            benchmarkMessage.append(Component.text("% CPU ", NamedTextColor.YELLOW));
            benchmarkMessage.append(Component.text(MathUtils.round(result.getUserPercentage(), 2), NamedTextColor.RED));
            benchmarkMessage.append(Component.text("% USER ", NamedTextColor.RED));
            benchmarkMessage.append(Component.text(MathUtils.round(result.getBlockedPercentage(), 2), NamedTextColor.LIGHT_PURPLE));
            benchmarkMessage.append(Component.text("% BLOCKED ", NamedTextColor.LIGHT_PURPLE));
            benchmarkMessage.append(Component.text(MathUtils.round(result.getWaitedPercentage(), 2), NamedTextColor.GREEN));
            benchmarkMessage.append(Component.text("% WAITED ", NamedTextColor.GREEN));
            benchmarkMessage.append(Component.newline());
        }
        return benchmarkMessage.build();
    }

    private void refreshData() {
        ThreadInfo[] threadInfo = THREAD_MX_BEAN.getThreadInfo(THREAD_MX_BEAN.getAllThreadIds());
        for (ThreadInfo threadInfo2 : threadInfo) {
            if (threadInfo2 == null) continue; // Can happen if the thread does not exist
            final String name = threadInfo2.getThreadName();
            if (THREADS.stream().noneMatch(name::startsWith)) continue;

            final long id = threadInfo2.getThreadId();

            final long lastCpuTime = lastCpuTimeMap.getOrDefault(id, 0L);
            final long lastUserTime = lastUserTimeMap.getOrDefault(id, 0L);
            final long lastWaitedTime = lastWaitedMap.getOrDefault(id, 0L);
            final long lastBlockedTime = lastBlockedMap.getOrDefault(id, 0L);

            final long blockedTime = threadInfo2.getBlockedTime();
            final long waitedTime = threadInfo2.getWaitedTime();
            final long cpuTime = THREAD_MX_BEAN.getThreadCpuTime(id);
            final long userTime = THREAD_MX_BEAN.getThreadUserTime(id);

            lastCpuTimeMap.put(id, cpuTime);
            lastUserTimeMap.put(id, userTime);
            lastWaitedMap.put(id, waitedTime);
            lastBlockedMap.put(id, blockedTime);

            final double totalCpuTime = (double) (cpuTime - lastCpuTime) / 1000000D;
            final double totalUserTime = (double) (userTime - lastUserTime) / 1000000D;
            final long totalBlocked = blockedTime - lastBlockedTime;
            final long totalWaited = waitedTime - lastWaitedTime;

            final double cpuPercentage = totalCpuTime / (double) time * 100L;
            final double userPercentage = totalUserTime / (double) time * 100L;
            final double waitedPercentage = totalWaited / (double) time * 100L;
            final double blockedPercentage = totalBlocked / (double) time * 100L;

            ThreadResult threadResult = new ThreadResult(cpuPercentage, userPercentage, waitedPercentage, blockedPercentage);
            resultMap.put(name, threadResult);
        }
    }
}

File: src/main/java/net/minestom/server/monitoring/BenchmarkManager.kt
package net.minestom.server.monitoring

import it.unimi.dsi.fastutil.longs.Long2LongMap
import net.minestom.server.utils.validate.Check.stateCondition
import net.minestom.server.MinecraftServer.Companion.exceptionManager
import it.unimi.dsi.fastutil.longs.Long2LongOpenHashMap
import net.kyori.adventure.text.Component
import net.minestom.server.monitoring.ThreadResult
import java.util.concurrent.ConcurrentHashMap
import net.minestom.server.monitoring.BenchmarkManager
import java.lang.Runnable
import java.lang.InterruptedException
import net.minestom.server.MinecraftServer
import net.kyori.adventure.text.format.NamedTextColor
import net.minestom.server.utils.MathUtils
import org.slf4j.LoggerFactory
import java.lang.management.ManagementFactory
import java.time.Duration
import java.util.*

/**
 * Small monitoring tools that can be used to check the current memory usage and Minestom threads CPU usage.
 *
 *
 * Needs to be enabled with [.enable]. Memory can then be accessed with [.getUsedMemory]
 * and the CPUs usage with [.getResultMap] or [.getCpuMonitoringMessage].
 *
 *
 * Be aware that this is not the most accurate method, you should use a proper java profiler depending on your needs.
 */
class BenchmarkManager {
    private val lastCpuTimeMap: Long2LongMap = Long2LongOpenHashMap()
    private val lastUserTimeMap: Long2LongMap = Long2LongOpenHashMap()
    private val lastWaitedMap: Long2LongMap = Long2LongOpenHashMap()
    private val lastBlockedMap: Long2LongMap = Long2LongOpenHashMap()
    private val resultMap: MutableMap<String, ThreadResult> = ConcurrentHashMap()
    private var enabled = false

    @Volatile
    private var stop = false
    private var time: Long = 0
    fun enable(duration: Duration) {
        stateCondition(enabled, "A benchmark is already running, please disable it first.")
        try {
            THREAD_MX_BEAN.isThreadContentionMonitoringEnabled = true
            THREAD_MX_BEAN.isThreadCpuTimeEnabled = true
        } catch (e: Throwable) {
            // Likely unsupported by the JVM (e.g. Substrate VM)
            LOGGER.warn("Could not enable thread monitoring", e)
            return
        }
        time = duration.toMillis()
        val thread = Thread(null, {
            while (!stop) {
                refreshData()
                try {
                    Thread.sleep(time)
                } catch (e: InterruptedException) {
                    exceptionManager.handleException(e)
                }
            }
            stop = false
        }, MinecraftServer.THREAD_NAME_BENCHMARK)
        thread.isDaemon = true
        thread.start()
        enabled = true
    }

    fun disable() {
        stop = true
        enabled = false
    }

    fun addThreadMonitor(threadName: String) {
        THREADS.add(threadName)
    }

    /**
     * Gets the heap memory used by the server in bytes.
     *
     * @return the memory used by the server
     */
    val usedMemory: Long
        get() = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()

    fun getResultMap(): Map<String, ThreadResult> {
        return Collections.unmodifiableMap(resultMap)
    }

    val cpuMonitoringMessage: Component
        get() {
            if (!enabled) return Component.text("CPU monitoring is disabled")
            val benchmarkMessage = Component.text()
            for ((name, result) in resultMap) {
                benchmarkMessage.append(Component.text(name, NamedTextColor.GRAY))
                benchmarkMessage.append(Component.text(": "))
                benchmarkMessage.append(Component.text(MathUtils.round(result.cpuPercentage, 2), NamedTextColor.YELLOW))
                benchmarkMessage.append(Component.text("% CPU ", NamedTextColor.YELLOW))
                benchmarkMessage.append(Component.text(MathUtils.round(result.userPercentage, 2), NamedTextColor.RED))
                benchmarkMessage.append(Component.text("% USER ", NamedTextColor.RED))
                benchmarkMessage.append(
                    Component.text(
                        MathUtils.round(result.blockedPercentage, 2),
                        NamedTextColor.LIGHT_PURPLE
                    )
                )
                benchmarkMessage.append(Component.text("% BLOCKED ", NamedTextColor.LIGHT_PURPLE))
                benchmarkMessage.append(
                    Component.text(
                        MathUtils.round(result.waitedPercentage, 2),
                        NamedTextColor.GREEN
                    )
                )
                benchmarkMessage.append(Component.text("% WAITED ", NamedTextColor.GREEN))
                benchmarkMessage.append(Component.newline())
            }
            return benchmarkMessage.build()
        }

    private fun refreshData() {
        val threadInfo = THREAD_MX_BEAN.getThreadInfo(THREAD_MX_BEAN.allThreadIds)
        for (threadInfo2 in threadInfo) {
            if (threadInfo2 == null) continue  // Can happen if the thread does not exist
            val name = threadInfo2.threadName
            if (THREADS.stream().noneMatch { prefix: String? ->
                    name.startsWith(
                        prefix!!
                    )
                }) continue
            val id = threadInfo2.threadId
            val lastCpuTime = lastCpuTimeMap.getOrDefault(id, 0L)
            val lastUserTime = lastUserTimeMap.getOrDefault(id, 0L)
            val lastWaitedTime = lastWaitedMap.getOrDefault(id, 0L)
            val lastBlockedTime = lastBlockedMap.getOrDefault(id, 0L)
            val blockedTime = threadInfo2.blockedTime
            val waitedTime = threadInfo2.waitedTime
            val cpuTime = THREAD_MX_BEAN.getThreadCpuTime(id)
            val userTime = THREAD_MX_BEAN.getThreadUserTime(id)
            lastCpuTimeMap[id] = cpuTime
            lastUserTimeMap[id] = userTime
            lastWaitedMap[id] = waitedTime
            lastBlockedMap[id] = blockedTime
            val totalCpuTime = (cpuTime - lastCpuTime).toDouble() / 1000000.0
            val totalUserTime = (userTime - lastUserTime).toDouble() / 1000000.0
            val totalBlocked = blockedTime - lastBlockedTime
            val totalWaited = waitedTime - lastWaitedTime
            val cpuPercentage = totalCpuTime / time.toDouble() * 100L
            val userPercentage = totalUserTime / time.toDouble() * 100L
            val waitedPercentage = totalWaited / time.toDouble() * 100L
            val blockedPercentage = totalBlocked / time.toDouble() * 100L
            val threadResult = ThreadResult(cpuPercentage, userPercentage, waitedPercentage, blockedPercentage)
            resultMap[name] = threadResult
        }
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(BenchmarkManager::class.java)
        private val THREAD_MX_BEAN = ManagementFactory.getThreadMXBean()
        private val THREADS: MutableList<String> = ArrayList()

        init {
            THREADS.add(MinecraftServer.THREAD_NAME_TICK_SCHEDULER)
            THREADS.add(MinecraftServer.THREAD_NAME_TICK)
        }
    }
}